File: langs/typescript/src/output/serialize.ts
/**
 * Deterministic JSON serialization for candidates, critical decisions,
 * and critical constraints.
 *
 * Field order for candidates: alternativeIds, triangularValue, normalizedValue, score, rank.
 * Float representation: JS default Number.prototype.toString — matches Python repr(float).
 * Do NOT use .toFixed, .toPrecision, or any locale-aware formatting.
 */
import type { CandidateM } from '../models/candidate.js';
import type { CriticalDecisionM } from '../models/critical-decision.js';
import type { CriticalConstraintM } from '../models/critical-constraint.js';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function orderedStringify(obj: any): string {
  return JSON.stringify(obj, null, 2) + '\n';
}

export function serializeCandidates(candidates: CandidateM[]): string {
  const arr = candidates.map((c) => ({
    alternativeIds: [...c.alternativeIds],
    triangularValue: {
      lower: c.triangularValue.lower,
      modal: c.triangularValue.modal,
      upper: c.triangularValue.upper,
    },
    normalizedValue: {
      positive: c.normalizedValue.positive,
      average: c.normalizedValue.average,
      negative: c.normalizedValue.negative,
    },
    score: c.score,
    rank: c.rank,
  }));
  return orderedStringify(arr);
}

export function serializeCriticalDecisions(decisions: CriticalDecisionM[]): string {
  const arr = decisions.map((d) => ({
    decisionId: d.decisionId,
    triangularValue: {
      lower: d.triangularValue.lower,
      modal: d.triangularValue.modal,
      upper: d.triangularValue.upper,
    },
    normalizedValue: {
      positive: d.normalizedValue.positive,
      average: d.normalizedValue.average,
      negative: d.normalizedValue.negative,
    },
    score: d.score,
    rank: d.rank,
  }));
  return orderedStringify(arr);
}

export function serializeCriticalConstraints(constraints: CriticalConstraintM[]): string {
  const arr = constraints.map((c) => ({
    constraintIndex: c.constraintIndex,
    kind: c.kind,
    eliminated: c.eliminated,
    total: c.total,
    redundant: c.redundant,
  }));
  return orderedStringify(arr);
}

File: langs/python/src/guidearch/output/serialize.py
"""Deterministic JSON serialization for TOPSIS outputs.

Float formatting: Python's default repr (full precision).
Field ordering is stable (insertion order in dict literals).
"""

from __future__ import annotations

import json
from typing import Any

from guidearch.models.candidate import CandidateM
from guidearch.models.critical_constraint import CriticalConstraintM
from guidearch.models.critical_decision import CriticalDecisionM
from guidearch.models.normalized_fuzzy import NormalizedFuzzyM
from guidearch.models.triangular_fuzzy import TriangularFuzzyM


def _tfm_to_dict(t: TriangularFuzzyM) -> dict[str, Any]:
    return {"lower": t.lower, "modal": t.modal, "upper": t.upper}


def _nfm_to_dict(n: NormalizedFuzzyM) -> dict[str, Any]:
    return {"positive": n.positive, "average": n.average, "negative": n.negative}


def candidates_to_dict(candidates: tuple[CandidateM, ...]) -> list[dict[str, Any]]:
    """Serialize a tuple of CandidateM to a list of dicts."""
    return [
        {
            "alternativeIds": list(c.alternative_ids),
            "triangularValue": _tfm_to_dict(c.triangular_value),
            "normalizedValue": _nfm_to_dict(c.normalized_value),
            "score": c.score,
            "rank": c.rank,
        }
        for c in candidates
    ]


def critical_decisions_to_dict(
    decisions: tuple[CriticalDecisionM, ...],
) -> list[dict[str, Any]]:
    """Serialize a tuple of CriticalDecisionM to a list of dicts."""
    return [
        {
            "decisionId": d.decision_id,
            "triangularValue": _tfm_to_dict(d.triangular_value),
            "normalizedValue": _nfm_to_dict(d.normalized_value),
            "score": d.score,
            "rank": d.rank,
        }
        for d in decisions
    ]


def critical_constraints_to_dict(
    constraints: tuple[CriticalConstraintM, ...],
) -> list[dict[str, Any]]:
    """Serialize a tuple of CriticalConstraintM to a list of dicts."""
    return [
        {
            "constraintIndex": c.constraint_index,
            "kind": c.kind,
            "eliminated": c.eliminated,
            "total": c.total,
            "redundant": c.redundant,
        }
        for c in constraints
    ]


def to_json(obj: Any) -> str:
    """Pretty-print to deterministic JSON (indent=2, sorted_keys=False)."""
    return json.dumps(obj, indent=2, ensure_ascii=False)
